enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,298 @@ package hit_tracker_client
+
+	ticker  <-chan time.Time
+	quit    chan struct{}
+	stopped atomic.Bool
+
+	mu                     sync.Mutex
+	maxPendingHitsPerGroup int
+	hitsByGroup            map[groupID]*cacheHits
+
+	rpcTimeout time.Duration
+	client     hitpb.HitTrackerServiceClient
+}
+
+func (h *HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	// For Action Cache hit-tracking, explicitly ignore everything. These cache
+	// artifacts should always be served from the authoritative cache which
+	// will take care of hit-tracking.
+	return &NoOpHitTracker{}
+}
+
+func (h *HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	// For CAS hit-tracking, use a hit-tracker that sends information about
+	// local cache hits to the RPC service at the configured backend.
+	return &HitTrackerClient{ctx: ctx, enqueueFn: h.enqueue, client: h.client, requestMetadata: requestMetadata}
+}
+
+type NoOpHitTracker struct {
+}
+
+func (h *NoOpHitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+}
+
+func (h *NoOpHitTracker) TrackMiss(d *repb.Digest) error {
+	return nil
 }
 
-func Register(env *real_environment.RealEnv) {
-	env.SetHitTrackerFactory(HitTrackerFactory{})
+func (h *NoOpHitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-// TODO(iain): hook this up with a to-be-added RPC service in the app.
-type HitTracker struct {
+func (h *NoOpHitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+type NoOpTransferTimer struct {
 }
 
-func (h HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (t *NoOpTransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	return nil
+}
+
+func (t *NoOpTransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
+	return nil
 }
 
-func (h *HitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+type HitTrackerClient struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	client          hitpb.HitTrackerServiceClient
+	requestMetadata *repb.RequestMetadata
 }
 
-func (h *HitTracker) TrackMiss(d *repb.Digest) error {
+// This should only be used for Action Cache hit tracking.
+func (h *HitTrackerClient) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+	alert.UnexpectedEvent("Unexpected call to SetExecutedActionMetadata")
+}
+
+// Misses hit the backing cache, so no need to report on hit-tracking.
```

Comment:
i understand the words in this comment but it leaves me with more questions than answers lol

Reply:
Haha sorry, I tried to reword it to be a bit more clear, but not sure if that helped or not?